### DISCUSSION TITLE
Binding `Version` object to `TextBox.Text` crashes the application.

### DISCUSSION THREAD

**Discussion by huaxing-yuan** [AvaloniaUI/Avalonia#8873] [Q&A] [answered]
When the user input in TextBox is not valid against a data type, most of time with Avalonia there will be an Error Message showing.
But When I want to bind a "Version" object to the TextBox.Text, user input will crash the application.

To reproduce the error: bind this sample model to a Window
```csharp
    public class Model
    {
        public int Int { get; set; } = 3;
        public double Double { get; set; } = 3;
        public DateTime DateTime { get; set; } = new DateTime(2022, 10, 2);
        public Version Version { get; set; } = new Version(1, 0);
        public string Name { get; set; } = "name";
    }
```

```xml
<TextBox Text="{Binding Version}" />
<TextBox Text="{Binding Int}" />
<TextBox Text="{Binding Name}" />
<TextBox Text="{Binding DateTime}" />
<TextBox Text="{Binding Double}" />
```

We can change any type with in invalid input, There will be an internal InvalidCastException and Error Message is showing as expected, but Application is not crashing.
<img width="331" alt="image" src="https://user-images.githubusercontent.com/37816968/187798758-e5b33e6b-1df0-438d-b2ab-99c1966f707b.png">

Now modify the Version, from default `1.0` to `1.1`. The moment when we delete the 0 (it leaves `1.` in the TextBox) the application crashes with FormatException.
Console Log:
```
Exception thrown: 'System.FormatException' in System.ComponentModel.TypeConverter.dll
An unhandled exception of type 'System.FormatException' occurred in System.ComponentModel.TypeConverter.dll
1. is not a valid value for Version.
```

My Question is:
Is there a proper way to safely convert `Version` to `string`, (or maybe some other types) with similar behavior ?

- I used Regular Expression Attribute with no lock. 
- using IValueConverter, I can ignore the error in ConvertBack, but there is no prompt of error, and previous valid value is lost; by throwing  InvalidCastException or DataValidationException, the app crashs.

Thanks for your help.

**Comment by rabbitism** [comment, marked as answer]
What about returning BindingNotification in converter?

**Reply by huaxing-yuan** [reply]
Wow, Thanks.
By returning a `BindingNotification` rather than throwing another Exception in the Converter, It worked !

**Reply by timunie** [reply]
_My Comment was not useful here, so deleted it._
